compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1892,11 +1892,11 @@ object Types {
     /** The signature of the last known denotation, or if there is none, the
      *  signature of the symbol
      */
-    protected def computeSignature(implicit ctx: Context): Signature = {
-      val lastd = lastDenotation
-      if (lastd != null) lastd.signature
-      else symbol.asSeenFrom(prefix).signature
-    }
+    protected def computeSignature(implicit ctx: Context): Signature =
+      lastDenotation match
+        case null => symbol.asSeenFrom(prefix).signature
+        case sd: SingleDenotation => sd.initial.signature
```

Comment:
Perhaps Denotation#signature itself should do the work of calling initial? That way we're sure to never use the wrong signature

Reply:
Case in point: the code just below for `currentSignature` calls `lastd.signature` and `lastd` might not be the initial signature. Similarly, `symbol.asSeenFrom(prefix).signature` above will also be computed at the current phase.

Reply:
As long as signatures are defined in denotations, I think they should apply to the owner denotation. If we wanted to always go for initial, then signatures should be defined on symbols. 

Agree that `currentSignature` needs to be fixed as well.

Reply:
> If we wanted to always go for initial, then signatures should be defined on symbols.

That might make sense, I can see how different prefixes can lead to different signatures for denotations, but I don't think that should ever be needed to disambiguate between overloads ?

Reply:
hmm I guess that would probably interfere with how we deal with merging denotations.

Reply:
Yes, I think it's better to leave them on denotations.